openmc/kinetics/solver.py

```diff
@@ -0,0 +1,831 @@
+        state_fwd = self.states['FORWARD_OUTER']
+        state_pre = self.states['PREVIOUS_OUTER']
+        iteration = 0
+
+        # Save the old power
+        power_old = state_fwd.power
+
+        while True:
+
+            if iteration == 0:
+                self.clock.times['FORWARD_OUTER'] = self.clock.times['FORWARD_OUTER'] +\
+                    self.clock.dt_outer
+                residual = np.inf
+            else:
+
+                # Take inner steps
+                for i in range(self.num_inner_time_steps):
+                    self.take_inner_step(i)
+
+                # Copy the shape, amp, and and precursors to FORWARD_OUTER
+                self.copy_states('FORWARD_INNER', 'FORWARD_OUTER')
+
+                new_power = state_fwd.power
+                residual_array = (power_old - new_power) / new_power
+                residual_array = openmc.kinetics.nan_inf_to_zero(residual_array)
+                num_fissile_regions = np.sum(power_old > 0.)
+                residual = np.sqrt((residual_array**2).sum() / num_fissile_regions)
+                power_old = new_power
+
+            if residual < self.outer_tolerance and iteration >= self.min_outer_iters:
+                print('  CONVERGED OUTER residual {}'.format(residual))
+                break
+            else:
+
+                print('UNCONVERGED OUTER residual {}'.format(residual))
+
+                # Increment outer iteration count
+                iteration += 1
+
+                # Run OpenMC on forward out state
+                if iteration == 1:
+                    self.run_openmc('FORWARD_OUTER')
+                else:
+                    self.run_openmc('FORWARD_OUTER', self.method)
+
+                state_fwd.extract_shape()
+
+                # Reset the inner states
+                self.copy_states('PREVIOUS_OUTER', 'FORWARD_INNER')
+                self.copy_states('PREVIOUS_OUTER', 'PREVIOUS_INNER')
+
+        # Dump data at FORWARD_OUT state to log file
+        state_fwd.dump_to_log_file
+
+        # Copy the converged forward state to the previous state
+        self.copy_states('FORWARD_OUTER', 'PREVIOUS_OUTER')
+
+        # Increment the clock time step
+        self.clock.outer_step += 1
+
+    def generate_tallies_file(self, time_point):
+
+        # Generate a new tallies file
+        tallies_file = openmc.Tallies()
+
+        # Get the MGXS library
+        mgxs_lib = self.states[time_point].mgxs_lib
+
+        # Add the tallies to the file
+        for mgxs in mgxs_lib.values():
+            tallies = mgxs.tallies.values()
+            for tally in tallies:
+                tallies_file.append(tally, True)
+
+        # Export the tallies file to xml
+        tallies_file.export_to_xml(self.directory + '/tallies.xml')
+
+    def solve(self):
+
+        # Create run directory
+        if not os.path.exists(self.directory):
+            os.makedirs(self.directory)
+
+        # Create states and run initial OpenMC on initial state
```

Comment:
To me it doesn't look like `create_state` runs OpenMC. does this comment need to be updated?